Added fpgaSolver, requires Xilinx Alveo U280 FPGA board

g-marchiori opened the pull request:
This PR contains the integration of a preconditioned ILU0-BiCGSTAB linear solver implementation which can be run on a Xilinx Alveo U280 FPGA board.

It is composed of two parts: the software backend that performs the ILU0 decomposition and communicates with the FPGA (included in this PR), and a _bitstream_ implementation part that is needed to configure the FPGA with the BiCGSTAB kernel functionality (this part has now been moved to the new repository https://github.com/OPM/FPGA ).

In order to compile the software part, included in this PR:
- the Xilinx runtime (XRT, v2.3) should be installed and setup (freely available at the [Xilinx](https://www.xilinx.com/support/download/index.html/content/xilinx/en/downloadNav/alveo/u280.html) website);
- `ENABLE_FPGA` must be set to `ON` when compiling with cmake, otherwise the FPGA backend compilation will be disabled.

To run flow using the fpgaSolver, a host with the Xilnx Alveo U280 board and already setup with the proper drivers and runtime is required. It's also possible to use a cloud-based platform that makes available the Xilinx Alveo U280 board.\
Note that this PR changes the command line parameter `--gpu-mode=[none|cusparse|opencl]` to a more generic `--accelerator-mode=[none|cusparse|opencl|fpga]`, in order to be consistent and to be able to add even more accelerator types in the future.\
The following command line parameters must be used:\
 `--matrix-add-well-contributions=true --accelerator-mode=fpga --fpga-bitstream=<bitstream_file>`\
Additionally, the reordering algorithm can be chosen by using the parameter:\
 `--opencl-ilu-reorder=[graph_coloring|level_scheduling]`\
which is currently set by default to _level scheduling_ for the FPGA because graph coloring tends to overflow the FPGA solver in many cases (e.g. it may generate too many colors).

Please note that **this kernel is still experimental**, and the following limitations apply:
- it may not work with some input cases (i.e. the system to be solved may overflow a kernel buffer, or the kernel may hang because of an internal error); it has been tested with some small cases (e.g. SPE1CASE1) and norne (NORNE_ATW2013, NORNE_ATW2013_1A_STDW);
- if the FPGA is selected, but the well contributions are not added, an error is printed and the program exits;
- only blocks of 3x3 are supported, and other block sizes will make flow fall back to DUNE.

Further documentation, including the building of the _bitstream_ part, will be soon added to the OPM wiki.

Additional notes about the files in this PR:
- some sources are still not compliant with the OPM coding rules (the code was originally C), they should be further cleaned up and ported to C++ syntax;
- the number of iterations and linearizations will change between executions with the same input data: this is due to the design of the FPGA BiCGSTAB kernel, where the order of the operations is affected by the memory latency on the various read ports of the kernel.

Further description of the kernel implementation and some performance results for the NORNE test case can be found in the paper ["Hardware Acceleration of HPC Computational Flow Dynamics using HBM-enabled FPGAs"](https://arxiv.org/abs/2101.01745).

bska commented:
I have not looked at the details of this work at all.  I'm just going to say that a 42,000 line, 129 file single commit patch is going to be very difficult to review in any sensible manner.  I do notice however that you introduce files which require Tcl and that you've implemented at least part of the build system in terms of GNU Make instead of OPM's CMake-based solution.  The latter, in particular, will make it harder to adopt this work into OPM.  Is there a reason that you chose to use Make over CMake?

joakim-hove commented:
I must echo the sentiment of @bska - the work might be pure gold, but for mere mortals it close to impossible to review and understand in any meaningful way. I will probably not be involved in reviewing this, but with the disclaimer that I have not studied the PR at all - and only have a very superficial knowledge of FPGA - here are some suggestion you can consider:

1. Start with an issue describing how the FPGA code will add to the general code. Will it be a pure add-on, or will it also affect the generic code? My hunch is that this is quite specialized code which will not see very much mainstream use, I would therefor suggest keeping FPGA changes as much as possible out of the generic code path.

2. A specific card model is mentioned - how generic is this FPGA code?

3. Describe the layout of code: how much is novel code written for OPM and how much is code inlined from existing external code? Are there new external dependencies from this?

4. How if the FPGA code related OpenCL?

Due to sheer size and complexity the current PR should in my opinion be closed, and then resubmitted as a series of smaller more manageable PR's. I would start with a PR containing only the external stuff along with the necessary build system changes. When that has been merged a followup PR with the changes for OPM could be submitted.

Good luck!

blattms commented:
Wow that is a massive contribution. Thanks.

I read your comment as 
- currently this cannot be built and used due to removed stuff.
- documentation on how to build the FPGA kernel is upcoming.

I agree that there needs to be more documentation on the build and the structure of the code. Otherwise it is hardly usable for others (which is the aim, I guess). I think this and the points above should be done before considering a merge.

We also should rethink whether all this code should live in opm-simulators. What I understood from private conversations with others is, that the fpga library is actually programming the fpga. Hence it needs to define all/most of the operations needed on the FPGA. This seems like something that could be used elsewhere apart form OPM or opm-simulators. It is also a very specialized library, where you need a lot of specialized domain knowledge to contribute. Have you considered moving that part to separate fpga-only module? Only if that is found and activated we would add the FPGA bindings for the bridge (which should be less lines that the OpenCL ones).

Considering the build systems. At least in opm-simulators the usual CMake procedure should be sufficient to build everything. With another module you might able to a different specialized build procedure if needed.

Anyway, I will try to review part of the code (starting with CMake and  the FPGA bindings for the bridge) on Monday or if lucky before.

BTW Are there any tests (did not really look for them, just wondering)?

blattms commented on CMakeLists.txt:
```diff
@@ -443,3 +473,13 @@ endif()
+  add_dependencies(opmsimulators FPGA_library)
+  target_link_libraries( opmsimulators PUBLIC ${_modules_dir}/opm-simulators/opm/simulators/linalg/bda/fpga/alveo_u280/vitis_20192/OPM-integration/fpga_lib_alveo_u280.a )
+  # link Xilinx-specific OpenCL library if OpenCL not found
+  if (NOT OpenCL_FOUND)
+    # used here is the default installation path for the library, change as needed
+    target_link_libraries( opmsimulators PUBLIC /opt/xilinx/xrt/lib/libxilinxopencl.so )
+  endif()
```
Can't we make CMake find the correct opencl implementation by setting the OCL_ROOT environment variable or OPENCL_INCLUDE_DIR OPENCL_LIBRARY as a user?

What you are doing might only work on your personal system.

g-marchiori replied:
After more testing, this library does not seem to be needed, hence the whole block has been removed.

blattms reviewed (changes requested):
Just some Cmake comments for now.

g-marchiori commented:
Thanks all for your comments.
Following internal discussions and the suggestions by @joakim-hove and @blattms, we're thinking about closing this PR and either split it or move it completely elsewhere; a decision will be made soon.

g-marchiori commented on CMakeLists.txt:
```diff
@@ -132,6 +132,35 @@ if(CUDA_FOUND)
 
+# include FPGA target only when CMAKE_ENABLE_FPGA is set to true
+if(CMAKE_ENABLE_FPGA)
+  find_file(CL_HPP CL/opencl.h)
+  if(NOT CL_HPP)
+    set(HAVE_FPGA 0)
+    message(WARNING " OpenCL headers were not found, deactivating FPGA target.")
```
Changed.
The file opencl.h is a wrapper that included OpenCL extensions, and is currently required to compile code that uses the Xilinx extensions.

g-marchiori commented on CMakeLists.txt:
```diff
@@ -132,6 +132,35 @@ if(CUDA_FOUND)
+    message(WARNING " OpenCL headers were not found, deactivating FPGA target.")
+  else()
+    set(HAVE_FPGA 1)
+    add_compile_options(-DPORTS_CONFIG=PORTS_2r_3r3w_ddr)  # selects the kernel's memory ports configuration; must be in sync with available kernel bitstream
```
Done.

g-marchiori commented on CMakeLists.txt:
```diff
@@ -132,6 +132,35 @@ if(CUDA_FOUND)
+    # add external project to compile the FPGA library
+    include (ExternalProject)
+    ExternalProject_Add(FPGA_library
+      SOURCE_DIR ${_modules_dir}/opm-simulators/opm/simulators/linalg/bda/fpga/alveo_u280/vitis_20192/OPM-integration
+      BUILD_IN_SOURCE 1
```
Removed and changed the makefile in order to create the outputs in the build directory.

g-marchiori commented on CMakeLists.txt:
```diff
@@ -132,6 +132,35 @@ if(CUDA_FOUND)
+      SOURCE_DIR ${_modules_dir}/opm-simulators/opm/simulators/linalg/bda/fpga/alveo_u280/vitis_20192/OPM-integration
+      BUILD_IN_SOURCE 1
+      # force the build step to always be run because source dependencies cannot be made explicit
+      BUILD_ALWAYS 1
```
This is needed to always execute the makefile, which in turn will trigger compilation only when any library file has been modified (thus it should not significantly impact the build time).
Since cmake does not know which files are used by the external project, this seems to be the only way to trigger an update if they are changed.

g-marchiori commented on CMakeLists.txt:
```diff
@@ -132,6 +132,35 @@ if(CUDA_FOUND)
+    # add external project to compile the FPGA library
+    include (ExternalProject)
+    ExternalProject_Add(FPGA_library
+      SOURCE_DIR ${_modules_dir}/opm-simulators/opm/simulators/linalg/bda/fpga/alveo_u280/vitis_20192/OPM-integration
```
Done.

g-marchiori commented:
The code in this PR has been now split into two parts: the FPGA solver integration will still live here, whereas the FPGA _bistream_ part (including the RTL source code and the C integration library) has been moved to a newly-created module (https://github.com/OPM/FPGA).
I changed the description of the to reflect the new situation.
I will also add documentation to the wiki after I'll have access to it.

Tongdongq commented:
I force pushed to squash a commit that removed tons of obsolete code with the initial commit, that new commit is now 48edd30. The commit that removed code was made because the new repo OPM/FPGA is created.
The latest commit on this branch (d9c7327) is tested on a U280 board with the masters of January 22, the results where as expected.

blattms commented:
jenkins build this please

blattms commented:
The FPGA module is that used as source only, or will it be compiled and a library used here?

blattms commented:
I'll just try to answer myself and keep on asking (probably dumb) questions:

Seems like there are to things build upstream in the makefile `$(KERNEL_PACKAGE) $(BINARY_CONTAINER_FILE)`. The odd thing is that KERNEL_PACKAGE needs to be set by the user. The other odd thing is that we are still building the FPGA module from here. I would have assumed that the modules is build stand alone  and later used here. Is there a reason not to do this?

g-marchiori commented:
The FPGA library from OPM/FPGA is currently compiled as external project in the opm-simulator's CMakeLists.txt.
By the way, the compilation of the FPGA-related code is enabled only when cmake is run with ENABLE_FPGA set to ON and the Xilinx XRT package is installed and setup.

g-marchiori commented:
> Seems like there are two things build upstream in the makefile: `$(KERNEL_PACKAGE) $(BINARY_CONTAINER_FILE)`. The odd thing is that KERNEL_PACKAGE needs to be set by the user before calling make there. The other odd thing is that we are still building the FPGA module from here. I would have assumed that the modules is built stand alone and later used here. Is there a reason not to do this?

If you're referring to the makefile.primary.vitis, that one is used to create the bitstream, not the software library.
And it needs to be called by using the makefile present in <FPGA module>/linearalgebra/ilu0bicgstab/xilinx/alveo_u280/vitis_20192/HW-implementation
The KERNEL_PACKAGE is a file that containt the RTL source code, while the BINARY_CONTAINER_FILE is the bitstream.
They're built separately also because the KERNEL_PACKAGE can be recreated in a matter of seconds (and used for development purposes), while the BINARY_CONTAINER_FILE takes hours to build.
The 
See the makefile in <FPGA git>/the makefile to be used

Tongdongq commented:
I turned the warnings into fatal_errors. When the user explicitly enables the FPGA, a silent warning is probably not enough.
The resulting binary would not have any FPGA code, which only becomes apparent when trying the use the FPGA at runtime.
The `fpga_lib_alveo_u280.a` library is created into opm-simulators/build/FPGA_library-prefix/*, it automatically detects changes to the source files in the FPGA repo. The FPGA repo stays unaffected by this building.

blattms commented on opm/simulators/linalg/bda/FPGABlockedMatrix.hpp:
```diff
@@ -0,0 +1,66 @@
+
+void blockSub(Block mat1, Block mat2, Block resMat);
+void blockMult(Block mat1, Block mat2, Block resMat);
+int blockInvert3x3(Block mat, Block res);
```
Why not reuse  [Opm::Detail::Inverter](/OPM/opm-simulators/blob/1af403c45d3bb317b5203d9fe3c937cf79680636/opm/simulators/linalg/MatrixBlock.hpp#L573)?

Tongdongq replied:
Reusing now

blattms reviewed (changes requested):
I tried reviewing this in one step, but am now realizing that this is quite and effort and I need a break. Hence regard this as a first part of the review.

Aprt from the individual comments, some more general comments
- There is a lot of plain C used. There are probably reasons for this. If there are please point them out with comments in the code. For cases where there aren't any, we need to (at least in the future) get rid of them eventually
- Maybe I am missing something and there is a real reason for this, but it seems like there is quite a bit of code duplication between the GPU OpenCL and the FPGA code, in particular
   - There now is BlockedMatrix and FPGABlockedMatrix. Aren't they very similar? Are both really needed? Using one would save quite some code.
   - There are now two functions to find the level scheduling (one in BILU0.hpp and one here). What is the difference? Can they be unified?
 
We really should prevent code duplication if possible.

Tongdongq commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -126,26 +126,27 @@ namespace Opm
                 const std::string opencl_ilu_reorder = EWOMS_GET_PARAM(TypeTag, std::string, OpenclIluReorder);
                 const int linear_solver_verbosity = parameters_.linear_solver_verbosity_;
-                bdaBridge.reset(new BdaBridge<Matrix, Vector, block_size>(gpu_mode, linear_solver_verbosity, maxit, tolerance, platformID, deviceID, opencl_ilu_reorder));
+                std::string fpga_bitstream = EWOMS_GET_PARAM(TypeTag, std::string, FpgaBitstream);
```
A user can in theory have multiple bitstreams for the same board, or multiple FPGA's on the same host. We could add a default location.

Tongdongq commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -242,46 +249,55 @@ namespace Opm
                         }
+                        if (use_fpga && accelerator_mode.compare("fpga") == 0) {
+                            OpmLog::warning("fpgaSolver did not converge, now trying Dune to solve current linear system...");
+                        }
```
8f34f7e

Tongdongq commented on opm/simulators/linalg/bda/FPGABILU0.cpp:
```diff
@@ -0,0 +1,540 @@
+            graph_coloring = true;
+        } else {
+            printf("ERROR: FPGABILU0: ilu reordering strategy not set correctly\n");
+            exit(1);
```
The ChowPatel merge introduced the unsupported `NONE` value to the enum.

Tongdongq commented on opm/simulators/linalg/bda/FPGABILU0.cpp:
```diff
@@ -0,0 +1,540 @@
+        delete[] diagIndex;
+        if (rowsPerColor != NULL) { free(rowsPerColor); }
+        if (LUMat != NULL) { if (LUMat->nnzValues != NULL) { free(LUMat->nnzValues); } }
+        if (LUMat != NULL) { free(LUMat); }
```
This is reduced now

Tongdongq commented on opm/simulators/linalg/bda/FPGABILU0.cpp:
```diff
@@ -0,0 +1,540 @@
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <stdio.h>
```
Header inclusion should be improved in 65a6776

Tongdongq commented on opm/simulators/linalg/bda/BdaBridge.cpp:
```diff
@@ -209,24 +255,74 @@ void BdaBridge<BridgeMatrix, BridgeVector, block_size>::solve_system(BridgeMatri
+        }
+
+        double *nnzsp = nullptr;
+        if (h_cols == nullptr) {
```
Code duplication reduced in c65b2f5.
The `h_rows` and `h_cols` vectors store the sparsity pattern in contiguous memory. This can be removed if there is a `Dune::BCRSMatrix` that guarantees the same thing.
For every linear solve, it is checked whether the sparsity pattern is stored already.

blattms replied:
Great. Thanks.

Tongdongq commented on opm/simulators/linalg/ISTLSolverEbos.hpp:
```diff
@@ -242,46 +249,55 @@ namespace Opm
                 }
 #endif
+                // WellContributions are not used when FPGA is enabled
+#if !HAVE_FPGA
```
Bug should be fixed.

Tongdongq commented:
I force-pushed again to add lots of small improvements, and remove code duplication.
I still have the older commits if needed.
Commit 65a6776, which reduces header inclusion in the `opm` namespace parts, is also in PR [3089](https://github.com/OPM/opm-simulators/pull/3089).

Tongdongq commented:
@blattms This PR is now ready for review, but I cannot find the button to get it out of draft mode. Maybe only Giacomo can do that.

blattms commented:
There are conflicts that need to be resolved. Probably due to merging the other PR

Tongdongq commented:
I rebased on top of a05bb1a, from 11 March 2021. All intermediary commits compile with the other masters of 2021-03-10 10:00.
The latest commit 8ea19c6 is benchmarked:
```
U280, NORNE_ATW_2013
858 (157+557+88+47), 1897, 1530, 21548 | 0
```
This time corresponds to earlier benchmarks.

blattms commented:
@ducbueno Any news about your board acquisition? Given that there might be major refactorings soon, we should probably merge this soon. Might be before you lay hands on any board, if that is not in the coming two weeks.

ducbueno commented:
@blattms the shipment of the board is scheduled for April 19. Besides that, there will be the setup time, which will probably take a while. You can go on with the merge.

blattms commented:
thanks for the prompt feedback. Will do.

blattms commented:
jenkins build this please
